### DISCUSSION TITLE
Cognitive relay networks using Sionna RTX

### DISCUSSION THREAD

**Discussion by yacine074** [NVlabs/sionna#116] [General]
Hello, 

I am interested in simulating a cognitive relay network that includes a primary transmitter (UP) and its destination (DP), a secondary full-duplex relay (R), and a secondary user (US) and its destination (DS) as described in this [paper](https://hal.science/hal-03534545/document). Is it possible to perform this network using Sionna RTX? 
Also, how can we simulate the relay's ability to transmit and receive simultaneously? And, is there a way to obtain the channel frequency response for only the line-of-sight (LoS) path?

PS: I am currently attempting to utilize the tutorial that has been published regarding [Sionna RTX](https://nvlabs.github.io/sionna/examples/Sionna_Ray_Tracing_Introduction.html).

Thanks in advance, and best regards.

**Comment by jhoydis** [comment]
Hi,

Yes, it is totally possible to simulate such a network in Sionna. It boils down to generating the channels between all devices and computing the transmitted and received signals. It seems that you have eight different channels in your problem. You can compute the outputs for these channels individually and then simply combine them in the way you want, e.g., to model the interference. You can also model a full-duplex relay with this approach. 

You can get the LoS component by keeping only the first tap of the channel impulse response (if you have a channel model with LoS).
If you want to use Sionna RT, you could simply set the ``max_depth`` parameters equal to one. This will only return the LoS component if any.

**Comment by Saman12191** [comment]
Hello,

I was wondering if there are any commands in Sionna RTX that can be used to differentiate between NLOS paths and LOS paths.

Thanks in advance

**Comment by faycalaa** [comment]
Hi,

You can use the (undocumented) ``objects`` field of the ``Paths`` object returned by ``compute_paths()``. It is an integer-valued tensor with shape ``[max_depth, num_targets, num_sources, max_num_paths]``.  It gives for every link the index of the object hit by the path for every interaction, up to the requested ``max_depth``.  -1 indicates that no object was hit. A LoS path has only -1 on the first dimension, i.e., does not hit anything.

**Comment by Saman12191** [comment]
Hi,

Thank you for your response. In my case, the paths.objects tensor is as follows:

<tf.Tensor: shape=(2, 1, 1, 8), dtype=int32, numpy=
array([[[[  -1, 2366,  323,  513, 2366,  513,  513, 1905]]],


       [[[  -1,   -1,   -1,   -1,  323,  323, 2366, 2366]]]], dtype=int32)>


Does this mean that the ray emitted by the transmitter initially hits the object with ID 2366, then encounters another object with the same ID (2366), and so on? What do the numbers following a -1 value represent in the case of a Line-of-Sight (LOS) path?

I'm planning to remove these LOS paths, but I'm unsure about the relationship between these numbers and the values of a and tau. 

Furthermore, I've observed that some object indices (such as 2366) are not present among the elements defined in the scene.objects list. How can I determine the names of these elements or the corresponding radio materials?

I would greatly appreciate any assistance you can provide.

Thank you.

**Comment by faycalaa** [comment]
The shape of this ``objects`` tensor is ``[max_depth = 2, num_targets = 1, num_sources = 1, max_num_paths = 8]``.

Therefore, a single link is simulated (``num_targets = num_sources = 1``), and for this link 8 paths were found (``max_num_paths = 8``) with up to 2 interactions for each path (``max_depth = 2``).

The first path does not intersect any object as the corresponding entry in the tensor is ``(-1, -1)`` for the ``max_depth`` dimension. This path is therefore a LoS path.

The next 3 paths hit a single object (``TX -> OBJECT -> RX``), with the id of the intersected object given by the first entry of the ``max_depth`` dimension (2366, 323, and 513, respectively). The second entry of this dimension is ``-1`` for these paths, which indicates that no other object was intersected.

The last 4 paths hit two objects (``TX -> OBJECT 1 -> OBJECT 2 -> RX``), and the pair of objects intersected is (2366, 323) for the first path, (513, 323) for the second path, and so on.

The ids of the objects correspond to the order in which they appear in ``scene.objects`` . Here is a code snippet you could use to get the name and material of the object from its id:

```
id = 2366
object = list(scene.objects.values())[id]
name = object.name
material = object.radio_material
```

**Comment by yacine074** [comment]
Hi,

Thank you so much for your response.

In our scenario, we have a single transmitter antenna and a single receiver antenna for both the primary and secondary network, and the channels are static. I would like to know if it is feasible to simulate a static channel using Sionna RTX. Additionally, I'm interested to understand if it is possible to model just one antenna instead of a grid or array of antennas (PlanarArray) in the simulation ?

Furthermore, I would appreciate your guidance regarding the most suitable type of antenna to consider for our work.

Thank you for your support and assistance. I truly appreciate it.

**Comment by faycalaa** [comment]
Hi,

The channel simulated by Sionna RT is static, i.e., the radio devices or objects in the scene are not moving. When computing the channel impulse responses (CIRs) from the paths using ``Paths2CIR``, you would need to set the velocities of the transmitters and receivers to zero.

To set a single antenna for the transmitter and/or receiver, you can define an array with a single element. For example, for setting a single vertically polarized isotropic antenna, set the ``tx_array`` and/or ``rx_array`` field of ``Scene`` to:

```
PlanarArray(num_rows=1,
            num_cols=1,
            vertical_spacing=0.5,
            horizontal_spacing=0.5,
            pattern="iso",
            polarization="V")
```
